### PR TITLE
Turns FileData into a C++-style class (phase 1 of #1369)

### PR DESCRIPTION
**Pull request by xsdg**
See #1369 for more context, but at a high level, this is a step towards being able to use C++-style fine-grained access control on the class methods, functions, and members to clearly define a public FileData API, as well as internal APIs between the different FileData subcomponents.  Furthermore, this will allow us to test the internal implementations (especially around refcounting) without making those internal implementations part of the public API.

### CODE REVIEW

**Comment by xsdg** [comment]
@qarkai @caclark I tried to include the minimum number of commits that still keeps mechanical changes separate from changes that actually need more thought to review.  Also, sadly, the 2-step copy procedure (from https://devblogs.microsoft.com/oldnewthing/20190919-00/?p=102904 ) is the only way I could find to retain file history across a copy, and it takes up 3 commits 😢

This is a big change, but I've tried to make sure everything is being run through tests and static analysis, and that the sequence of changes is fairly logical and easy to understand/review.

**Comment by caclark** [comment]
@xsdg - or anyone else...

Sorry, I have more problems with git. This pull request conflicts with the latest Russian translation.
What is the correct way to resolve the conflict?

Thanks

File: src/filedata.cc [hunk 1]
```diff
@@ -21,3561 +21,565 @@
...
-		*d = *s;
-		s++;
-		d++;
-		i--;
-		}
-	*d = '\0';
-
-	g_free(a);
-	return b;
+	return FileData::text_from_size(size);
```
Comment (qarkai) [inline src/filedata.cc:26]: Not sure `text_from_*` belongs to `FileData` but it's better to leave it there for now.

File: src/filedata.h [hunk 2]
```diff
@@ -132,8 +132,182 @@ struct FileData {
...
+
+#ifdef DEBUG_FILEDATA
+	FileData *file_data_ref(const gchar *file = __builtin_FILE(), gint line = __builtin_LINE());
+	void file_data_unref(const gchar *file = __builtin_FILE(), gint line = __builtin_LINE());
+#else
+	FileData *file_data_ref();
+	void file_data_unref();
+#endif
+
+	void file_data_lock(FileData *fd);
```
Comment (qarkai) [inline src/filedata.h:168]: Is removing `FileData` parameter from non-static methods planned in the next steps?

**Comment by qarkai** [comment]
> @xsdg - or anyone else...
> 
> Sorry, I have more problems with git. This pull request conflicts with the latest Russian translation. What is the correct way to resolve the conflict?
> 
> Thanks

Usually author of PR should resolve conflicts and update PR.

**Comment by caclark** [comment]
If I run `patch -p1 < 1375.diff` and then run `./scripts/test-all.sh` I get the error:

`../src/search.cc:3828:19: error: redundant explicit casting to the same type 'FileData *' as the sub-expression, remove this casting [readability-redundant-casting,-warnings-as-errors]`
` 3828 |                 file_data_unref((FileData *)mfd->fd);`
`      |                                 ^~~~~~~~~~~~`
`../src/search.cc:272:12: note: source type originates from referencing this member`
`  272 |         FileData *fd;`
`      |         ~~~~~~~~~~^`

I do not understand why the error did not show when the pull request tests ran on GitHub.

File: src/filedata.h (hunk 2 above)
Comment (xsdg) [reply src/filedata.h:168]: Yeah.  Going to drop that, and also drop the C-style "file_data_" name prefix.  So the end state for this method is `void lock();`

**Comment by xsdg** [comment]
> If I run `patch -p1 < 1375.diff` and then run `./scripts/test-all.sh` I get the error:
> 
> `../src/search.cc:3828:19: error: redundant explicit casting to the same type 'FileData *' as the sub-expression, remove this casting [readability-redundant-casting,-warnings-as-errors]` ` 3828 | file_data_unref((FileData *)mfd->fd);` ` | ^~~~~~~~~~~~` `../src/search.cc:272:12: note: source type originates from referencing this member` ` 272 | FileData *fd;` ` | ~~~~~~~~~~^`
> 
> I do not understand why the error did not show when the pull request tests ran on GitHub.

You're probably using a newer version of clang.  This check was added very recently: \
https://github.com/llvm/llvm-project/issues/67534

(Incidentally, this is one of the big downsides of things that behave like `-Wall` -- new compiler versions can cause code to stop compiling that used to compile without issue.)  My preference is to fix this in a separate PR.

File: src/filedata.cc (hunk 1 above)
Comment (xsdg) [reply src/filedata.cc:26]: Agreed.  I think there are a lot of util functions that have ended up scattered around the codebase, and refactors like this one will make it easier to spot them and to move them to more appropriate homes.  I think it'll be best to evaluate that once filedata.cc is split into separate files, though.

**Comment by xsdg** [comment]
Rebased to pick up ru.po update.  Should patch cleanly now.

**Comment by caclark** [comment]
Implemented in commits 43c3a767a97b 7a5da6e9232f22cf 0c26f3819bd97 e5648fc1579b555 9414f22efcafd3b 0832137ce92c5 4978b5bc6d3121d 1ebf122bce5e6 73765be18987d 13e8d9f487d9594 edf7b46fcd9ec ab671bdfb1bed 18183df6196d 535c3054a30f 0f50e13335

File: src/filedata.h [hunk 3]
```diff
@@ -74,7 +70,11 @@ struct FileDataChangeInfo {
 	gboolean regroup_when_finished;
 };
 
-struct FileData {
+class FileData {
```
Comment (qarkai) [inline src/filedata.h:73]: Reminder for the next PR. Convert forward declarations from `struct FileData` to `class FileData`.

**Pull request closed without merging**